### ISSUE TITLE
Group loses the kwargs

### ISSUE THREAD

**Issue by Mulugruntz** [celery/celery#3426]
Hi!
I've stumbled upon an issue that seems to be rather blocking (or at least troublesome, as I have an extremely dirty workaround).

In the docs ( http://docs.celeryproject.org/en/latest/userguide/canvas.html#partials ) it says :

> Any keyword arguments added will be merged with the kwargs in the signature, with the new keyword arguments taking precedence:
```python
>>> s = add.s(2, 2)
>>> s.delay(debug=True)                    # -> add(2, 2, debug=True)
>>> s.apply_async(kwargs={'debug': True})  # same
```

So apparently, it is okay to give kwargs to groups. However, when tried, this raises a TypeError, as the kwargs are eaten up and never reach the function:

```python
import sys
print sys.version
# 2.7.12 (default, Jul  1 2016, 15:12:24) 
# [GCC 5.4.0 20160609]
import celery, kombu, vine, amqp, billiard
print 'celery', celery.__version__, 'kombu', kombu.__version__, 'vine', vine.__version__, 'amqp', amqp.__version__, 'billiard', billiard.__version__
# celery 4.0.0rc3 kombu 4.0.0rc3 vine 1.1.1 amqp 2.0.3 billiard 3.5.0.0

######

import celery

celery.current_app.conf.update(
    CELERY_ALWAYS_EAGER=True,
)

@celery.task
def add(x, y):
    return x+y

def farg1(): return [add(1, 1), add(1, 1)]
def farg2(): return celery.group([add.s(), add.s()]).apply(args=(1, 1)).get()
def farg3(): return celery.group([add.s(), add.s()]).apply_async(args=(1, 1)).get()
def farg4(): return celery.group([add.s(), add.s()]).delay(1, 1).get()
def fkwarg1(): return [add(x=1, y=1), add(x=1, y=1)]
def fkwarg2(): return celery.group([add.s(), add.s()]).apply(kwargs=dict(x=1, y=1)).get()
def fkwarg3(): return celery.group([add.s(), add.s()]).apply_async(kwargs=dict(x=1, y=1)).get()
def fkwarg4(): return celery.group([add.s(), add.s()]).delay(x=1, y=1).get()


for fun in (farg1, farg2, farg3, farg4, fkwarg1, fkwarg2, fkwarg3, fkwarg4):
	try:
		fun()
	except TypeError, e:
		print fun.__name__, e
	else:
		print fun.__name__, 'OK'

# farg1 OK
# farg2 OK
# farg3 OK
# farg4 OK
# fkwarg1 OK
# fkwarg2 add() takes exactly 2 arguments (0 given)
# fkwarg3 add() takes exactly 2 arguments (0 given)
# fkwarg4 add() takes exactly 2 arguments (0 given)

```

The issue seems to be in https://github.com/celery/celery/blob/master/celery/canvas.py#L985
As we can see, kwargs is never used. Plain ignored.

In self._prepared(), args are used up and when in the sig.apply(), they reappear as they are merged with emptiness. Yep, we don't give anything to sig.apply(), besides options.

So basically, when a task is completely partial, the later args are injected as if they were partial args, because there is no real way to merge the partial args/kwargs with the later args/kwargs. Well, there is a way but it's not currently used.

My fix:
```python
    def apply(self, args=(), kwargs={}, **options):
        app = self.app
        if not self.tasks:
            return self.freeze()  # empty group returns GroupResult
        options, group_id, root_id = self._freeze_gid(options)
#################### Changed args into []
        tasks = self._prepared(self.tasks, [], group_id, root_id, app)
        return app.GroupResult(group_id, [
#################### Added args=args, kwargs=kwargs
            sig.apply(args=args, kwargs=kwargs, **options) for sig, _ in tasks
        ])
```
So, I removed the args from the _prepared and added args and kwargs to sig.apply.
It works fine. I've tested it with no partials, with partials, with overridden partials. All cases work.

However, as one might notice, I've used CELERY_ALWAYS_EAGER=True for my tests. So, it fixes it when eager, but probably not in async.

As we can see ( https://github.com/celery/celery/blob/master/celery/canvas.py#L956 ), in apply_async, kwargs is only used in eager mode, when it falls back to apply().
So this is still to be fixed.
I've done a fix for that as well, but I haven't tested it yet:

```python

#################### Added in argument list : args=None, kwargs=None 
    def _apply_tasks(self, tasks, producer=None, app=None, p=None,
                     add_to_parent=None, chord=None, args=None, kwargs=None, **options):
        app = app or self.app
        with app.producer_or_acquire(producer) as producer:
            for sig, res in tasks:
                sig.apply_async(producer=producer, add_to_parent=False,
                                chord=sig.options.get('chord') or chord,
#################### Added this line:
                                args=args, kwargs=kwargs,
                                **options)

                # adding callback to result, such that it will gradually
                # fulfill the barrier.
                #
                # Using barrier.add would use result.then, but we need
                # to add the weak argument here to only create a weak
                # reference to the object.
                if p and not p.cancelled and not p.ready:
                    p.size += 1
                    res.then(p, weak=True)
                yield res  # <-- r.parent, etc set in the frozen result.

    def apply_async(self, args=(), kwargs=None, add_to_parent=True,
                    producer=None, **options):
        app = self.app
        if app.conf.task_always_eager:
            return self.apply(args, kwargs, **options)
        if not self.tasks:
            return self.freeze()

        options, group_id, root_id = self._freeze_gid(options)
#################### Changed args into []
        tasks = self._prepared(self.tasks, [], group_id, root_id, app)
        p = barrier()
#################### Added args=args, kwargs=kwargs
        results = list(self._apply_tasks(tasks, producer, app, p, args=args, kwargs=kwargs, **options))
        result = self.app.GroupResult(group_id, results, ready_barrier=p)
        p.finalize()

        # - Special case of group(A.s() | group(B.s(), C.s()))
        # That is, group with single item that is a chain but the
        # last task in that chain is a group.
        #
        # We cannot actually support arbitrary GroupResults in chains,
        # but this special case we can.
        if len(result) == 1 and isinstance(result[0], GroupResult):
            result = result[0]

        parent_task = app.current_worker_task
        if add_to_parent and parent_task:
            parent_task.add_trail(result)
        return result
```

So, there it is!

Moreover, there might be the same issue with chain ( https://github.com/celery/celery/blob/master/celery/canvas.py#L715 ) as kwargs are eaten up as well. I haven't tested with chain though.

Regards,
Mulugruntz

**Comment by Mulugruntz** [comment]
Okay, I've done a patcher, for the people who would want to fix this issue without waiting for a released fix. Or even to test a fix before releasing it.

```python

class Patcher(object):
    def __init__(self, canvas):
        self.__canvas = canvas
        self.__saved_apply = canvas.group.apply
        self.__saved_apply_async = canvas.group.apply_async
        self.__saved__apply_tasks = canvas.group._apply_tasks

        from vine import barrier
        from celery.result import GroupResult

        def _apply_tasks(self, tasks, producer=None, app=None, p=None,
                         add_to_parent=None, chord=None, args=None, kwargs=None, **options):
            app = app or self.app
            with app.producer_or_acquire(producer) as producer:
                for sig, res in tasks:
                    sig.apply_async(producer=producer, add_to_parent=False,
                                    chord=sig.options.get('chord') or chord,
                                    args=args, kwargs=kwargs,
                                    **options)

                    # adding callback to result, such that it will gradually
                    # fulfill the barrier.
                    #
                    # Using barrier.add would use result.then, but we need
                    # to add the weak argument here to only create a weak
                    # reference to the object.
                    if p and not p.cancelled and not p.ready:
                        p.size += 1
                        res.then(p, weak=True)
                    yield res  # <-- r.parent, etc set in the frozen result.

        def apply_async(self, args=(), kwargs=None, add_to_parent=True,
                        producer=None, **options):
            app = self.app
            if app.conf.task_always_eager:
                return self.apply(args, kwargs, **options)
            if not self.tasks:
                return self.freeze()

            options, group_id, root_id = self._freeze_gid(options)
            tasks = self._prepared(self.tasks, [], group_id, root_id, app)
            p = barrier()
            results = list(self._apply_tasks(tasks, producer, app, p, args=args, kwargs=kwargs, **options))
            result = self.app.GroupResult(group_id, results, ready_barrier=p)
            p.finalize()

            # - Special case of group(A.s() | group(B.s(), C.s()))
            # That is, group with single item that is a chain but the
            # last task in that chain is a group.
            #
            # We cannot actually support arbitrary GroupResults in chains,
            # but this special case we can.
            if len(result) == 1 and isinstance(result[0], GroupResult):
                result = result[0]

            parent_task = app.current_worker_task
            if add_to_parent and parent_task:
                parent_task.add_trail(result)
            return result

        def apply(self, args=(), kwargs={}, **options):
            app = self.app
            if not self.tasks:
                return self.freeze()  # empty group returns GroupResult
            options, group_id, root_id = self._freeze_gid(options)
            tasks = self._prepared(self.tasks, [], group_id, root_id, app)
            return app.GroupResult(group_id, [
                sig.apply(args=args, kwargs=kwargs, **options) for sig, _ in tasks
                ])

        self.__patched_apply = apply
        self.__patched_apply_async = apply_async
        self.__patched__apply_tasks = _apply_tasks

    def patch(self, do_it=True):
        self._patch() if do_it else self._unpatch()

    def _patch(self):
        self.__canvas.group.apply = self.__patched_apply
        self.__canvas.group.apply_async = self.__patched_apply_async
        self.__canvas.group._apply_tasks = self.__patched__apply_tasks

    def _unpatch(self):
        self.__canvas.group.apply = self.__saved_apply
        self.__canvas.group.apply_async = self.__saved_apply_async
        self.__canvas.group._apply_tasks = self.__saved__apply_tasks

# Then, let's use it
import celery.canvas
p = Patcher(celery.canvas)
p.patch(True)  # Apply the patch
# do your stuff
p.patch(False)  # Remove the patch
```

Enjoy!

**Comment by ask** [comment]
Hey! Thanks for reporting this. Could you submit a pull request against master with this?  I can probably convert this into a patch, but I would like you to retain attribution of the fix.

**Comment by Mulugruntz** [comment]
Hi! Sure I can do that.

**Comment by ask** [comment]
Closed by #3426

**Comment by ask** [comment]
urm, that is: closed by #3433
